args <- commandArgs(trailingOnly = T)
# args <- c(11,0.1,0.9)
args <- as.numeric(args)
comp <- as.integer(args[1]):(as.integer(args[1])+4)
cut_pct <- args[2]

# noise level in y
alpha_y = (1-args[3])/args[3]

print('components')
print(comp)
print('Causal percentage')
print(cut_pct)
print('Heritability of y')
print(args[3])

arrID <- as.numeric(Sys.getenv('SLURM_ARRAY_TASK_ID'))
set.seed(arrID)

# suffix of file name generated by this file
name_suf <- paste('_', comp[1],'_',cut_pct,'_',args[3],'_snp','_',arrID, sep = '')

library(dplyr)
library(magrittr)
library(data.table)
library(OmicsPLS)
library(DescTools)
library(ggplot2)
library(pROC)
library(PO2PLS)
library(GenABEL)
# library(MultiABEL)
# library(lassosum)

#######################################################################
# # load X
# x <- readRDS(paste('./dat_rep/dat_training_',arrID,'.rds', sep = ''))
# x_t <- readRDS(paste('./dat_rep/dat_test_',arrID,'.rds', sep = ''))
# dim(x); dim(x_t)
# 
# # For PO2PLS
# gene <- readRDS(paste('./dat_rep/GeneData_training_',arrID,'.rds', sep = ''))
# gene %>% dim
# gene_t <- readRDS(paste('./dat_rep/GeneData_test_',arrID,'.rds', sep = ''))
# 
# PCA_weight <- readRDS(paste('./dat_rep/PCAweight_',arrID,'.rds', sep = ''))
# # sapply(PCA_weight, ncol) %>% sum
# # sapply(PCA_weight, nrow) %>% sum
# w <- readRDS('w_cut_snp.rds')
# ###############
# 
# # Generate 5 scores based on glycan groups
# T_q <- matrix(NA, 2000, 5)
# T_q_t <- matrix(NA, 1465, 5)
# 
# C <- cbind(rep(1,23), # all
#            c(rep(0,5), rep(1,5), rep(0,4), 1, rep(1,8)), # Monogalactosylation
#            c(rep(0,10), rep(1,4), 0, rep(1,3), 0, rep(1,4)), # Digalactosylation
#            c(1,0,1,0,1,0,rep(1,4),0,0,1,1,1,0,1,1,0,0,0,1,1), #fucosylation
#            c(rep(0,4), 1, rep(0,3), 1,1,0,1,0,1,rep(0,3),1,rep(0,4),1)) # bisectingGlcNAc
# 
# # cut off loadings
# w <- w[[as.character(cut_pct)]][,comp]
# 
# if(all.equal(rownames(w), colnames(x))!=T) stop('snps and loadings do not match')
# 
# # Get T
# T_q <- x %*% w
# T_q_t <- x_t %*% w
# 
# 
# #######################################################################
# # generate Y
# y_nonoise <- rbind(T_q,T_q_t) %*% t(C)
# 
# sd_y <- apply(y_nonoise, 2, sd)
# 
# ff <- scale(matrix(rnorm(3465*23,0,1),3465,23)) %*% diag(sd_y) * sqrt(alpha_y)
# y_all <- y_nonoise + ff
# y <- y_all[1:2000,]
# y_t <- y_all[-(1:2000),]
# 
# ###########
# # for gwas
# y_gwas <- y
# colnames(y_gwas) <- paste('y',1:23, sep = '')
# y_gwas <- cbind(id=rownames(gene), sex=rep(0,2000), y_gwas) # replace sex here
# write.table(y_gwas, file=paste('./fit_dat/y_gwas',name_suf, sep = ''), col.names =T, quote = F, row.names = F)
# 
# # for PRSice-2
# y_prsice <- y_gwas[,-2]
# colnames(y_prsice)[1] <- 'IID'
# write.table(y_prsice, file=paste('./fit_dat/y_prsice', name_suf, sep = ''), col.names =T, quote = F, row.names = F)
# 
# #######################################################################
# # generate Z
# 
# # Z from PCA of true Y
# z_true <- prcomp(y_nonoise)$x[,1]
# ###############################################
# # Z1 with 90% of heritability
# alpha_z1 <- (1-0.9)/0.9
# z1_all <- z_true + rnorm(3465, 0, sd(z_true)*sqrt(alpha_z1))
# z1 <- z1_all[1:2000]
# z1_t <- z1_all[-(1:2000)]
# # z as liability with prevalence of 0.3
# z1_bi <- ifelse(z1>quantile(z1,0.7),1,0)
# z1_bi_t <- ifelse(z1_t>quantile(z1_t,0.7),1,0)
# 
# ###############################################
# # z2 with 50% of heritability
# alpha_z2 <- (1-0.5)/0.5
# z2_all <- z_true + rnorm(3465, 0, sd(z_true)*sqrt(alpha_z2))
# z2 <- z2_all[1:2000]
# z2_t <- z2_all[-(1:2000)]
# # z as liability with prevalence of 0.3
# z2_bi <- ifelse(z2>quantile(z2,0.7),1,0)
# z2_bi_t <- ifelse(z2_t>quantile(z2_t,0.7),1,0)
# ###############################################
# # for gwas
# z_gwas <- cbind(z1=z1, z2=z2)
# z_gwas <- cbind(id=rownames(gene), sex=rep(0,2000), z_gwas)
# write.table(z_gwas, file=paste('./fit_dat/z_gwas',name_suf, sep = ''), col.names =T, quote = F, row.names = F)
# 
# # for PRSice-2
# z_prsice <- z_gwas[,-2]
# colnames(z_prsice)[1] <- 'IID'
# write.table(z_prsice, file=paste('./fit_dat/z_prsice', name_suf, sep = ''), col.names =T, quote = F, row.names = F)
# ###############################################
# 
# save(T_q, T_q_t, w, y_nonoise, y, y_t, z_true, z1, z1_t, z1_bi, z1_bi_t,
#      z2, z2_t, z2_bi, z2_bi_t, file = paste('./fit_dat/dat_yz', name_suf, '.RData', sep = ''))
# #######################################################################
# # # model fitting
# r=5
# rx=5
# ry=0
# 
# # # PLS
# # fit_pls <- mixOmics::pls(x,y,ncomp = r, scale = F, mode = "canonical")
# # print('PLS: correlation between T and U')
# # cor(fit_pls$variates$X, fit_pls$variates$Y)
# 
# # O2PLS
# fit_o2 <- o2m(gene,y,r,rx,ry)
# summary(fit_o2)
# saveRDS(fit_o2, file=paste('./fit_files/fit_o2', name_suf, '.rds', sep = ''))
# 
# # O2PLS
# fit_o2snp <- o2m(x,y,r,rx,ry)
# summary(fit_o2snp)
# saveRDS(fit_o2snp, file=paste('./fit_files/fit_o2snp', name_suf, '.rds', sep = ''))
# # print('O2PLS SNP: correlation between T and U')
# # cor(fit_o2snp$Tt,fit_o2snp$U)
# 
# # # SO2PLS
# # # fit_so2 <- o2m(x,y,r,rx,ry, sparsity = T, keepx = 70)
# # fit_so2 <- o2m(x,y,r,rx,ry, sparsity = T, keepx = 200)
# # summary(fit_so2)
# # print('SO2PLS: correlation between T and U')
# # cor(fit_so2$Tt,fit_so2$U)
# 
# #PO2PLS
# fit_p <- PO2PLS(gene,y,r,rx,ry, steps = 1000, tol=0.01, init_param = 'o2m')
# summary(fit_p)
# saveRDS(fit_p, file=paste('./fit_files/fit_p', name_suf, '.rds', sep = ''))
# 
# print('PO2PLS: correlation between T and U')
# with(fit_p$latent_vars, cor(mu_T, mu_U))
# 
# sd_B <- variances_inner.po2m(fit_p, gene, y)
# # fit_p$parameters$B; sd_B
# z_B <- (fit_p$parameters$B %>% diag) / sd_B
# p_B <- 2*(1-pnorm(z_B))
# # p-value of B
# print("p-value of global test")
# print(p_B)


##################################
# PO2PLS with 10 X-specific component
gene <- readRDS(paste('./dat_rep/GeneData_training_',arrID,'.rds', sep = ''))
load(file=paste('./fit_dat/dat_yz', name_suf, '.RData', sep = ''))

# PO2PLS
fit_pex <- PO2PLS(gene,y,5,15,0, steps = 1500, tol=0.01, init_param = 'o2m')
summary(fit_pex)
saveRDS(fit_pex, file=paste('./fit_files/fit_pex15', name_suf, '.rds', sep = ''))
##################################
# 
# # GWAS Y
# # convert.snp.ped('chr1.ped','chr1.map', outfile = 'chr1_ABEL.raw', mapHasHeaderLine = F)
# gwaa_data <- load.gwaa.data(phenofile = paste('./fit_dat/y_gwas',name_suf, sep = ''),
#                             genofile = paste('./dat_rep/ABEL_',arrID,'.raw', sep = ''),
#                             force = F, makemap = FALSE, sort = F, id = "id")
# # fix position
# gwaa_data@gtdata@map <- read.table(paste('./dat_rep/chr_all_training_',arrID,'.map', sep = ''), header = F)[,4]
# # gwaa_data@phdata %>% head
# 
# for(yi in 1:23){
#   uni <- qtscore(formula(paste('y',yi,'~sex',sep = '')), data=gwaa_data,
#                  trait.type = "gaussian", times = 1, quiet = FALSE,
#                  bcast = 10, clambda = TRUE, propPs = 1, details = TRUE)
# 
#   sum_stats <- descriptives.scan(uni, sortby = 'P2df', top = uni@dimresults[1])
#   sum_stats %<>%  mutate(SNP=rownames(sum_stats))
#   # # A2 is effective
#   # effallele(uni)
#   # all.equal(uni@annotation$A2, effallele(uni))
# 
#   write.table(sum_stats, file=paste('./fit_files/GWAS_y',yi, name_suf, '.txt', sep = ''), row.names = F, col.names = T, quote = F)
# }
# 
# 
# # GWAS Z
# gwaa_data <- load.gwaa.data(phenofile = paste('./fit_dat/z_gwas',name_suf, sep = ''),
#                             genofile = paste('./dat_rep/ABEL_',arrID,'.raw', sep = ''),
#                             force = F, makemap = FALSE, sort = F, id = "id")
# # fix position
# gwaa_data@gtdata@map <- read.table(paste('./dat_rep/chr_all_training_',arrID,'.map', sep = ''), header = F)[,4]
# # gwaa_data@phdata %>% head
# 
# for(zi in 1:2){
#   uni <- qtscore(formula(paste('z',zi,'~sex',sep = '')), data=gwaa_data,
#                  trait.type = "gaussian", times = 1, quiet = FALSE,
#                  bcast = 10, clambda = TRUE, propPs = 1, details = TRUE)
# 
#   sum_stats <- descriptives.scan(uni, sortby = 'P2df', top = uni@dimresults[1])
#   sum_stats %<>%  mutate(SNP=rownames(sum_stats))
#   # # A2 is effective
#   # effallele(uni)
#   # all.equal(uni@annotation$A2, effallele(uni))
# 
#   write.table(sum_stats, file=paste('./fit_files/GWAS_z',zi, name_suf, '.txt', sep = ''), row.names = F, col.names = T, quote = F)
# }


print('training R done')