arrID <- as.numeric(Sys.getenv('SLURM_ARRAY_TASK_ID'))
set.seed(arrID)

# suffix of file name generated by this file
name_suf <- paste0('_simu2_', arrID)

library(dplyr)
library(magrittr)
library(data.table)
library(OmicsPLS)
library(DescTools)
library(ggplot2)
library(pROC)
library(PO2PLS)
library(GenABEL)
# library(MultiABEL)
# library(lassosum)

#######################################################################
# # load X
# x <- readRDS(paste('./dat_rep/dat_training_',arrID,'.rds', sep = ''))
# x_t <- readRDS(paste('./dat_rep/dat_test_',arrID,'.rds', sep = ''))
# dim(x); dim(x_t)
# 
# # For PO2PLS
# gene <- readRDS(paste('./dat_rep/GeneData_training_',arrID,'.rds', sep = ''))
# gene %>% dim
# gene_t <- readRDS(paste('./dat_rep/GeneData_test_',arrID,'.rds', sep = ''))
# 
# PCA_weight <- readRDS(paste('./dat_rep/PCAweight_',arrID,'.rds', sep = ''))
# # sapply(PCA_weight, ncol) %>% sum
# # sapply(PCA_weight, nrow) %>% sum
# w <- readRDS('w_cut_gene.rds')
# ###############
# 
# # Generate 5 scores based on glycan groups
# 
# C1 <- cbind(rep(1,23), # all
#            c(rep(0,5), rep(1,5), rep(0,4), 1, rep(1,8)), # Monogalactosylation
#            c(rep(0,10), rep(1,4), 0, rep(1,3), 0, rep(1,4)), # Digalactosylation
#            c(1,0,1,0,1,0,rep(1,4),0,0,1,1,1,0,1,1,0,0,0,1,1), #fucosylation
#            c(rep(0,4), 1, rep(0,3), 1,1,0,1,0,1,rep(0,3),1,rep(0,4),1)) # bisectingGlcNAc
# 
# C2 <- cbind(c(rep(1,5),0,rep(1,4)), # chylomicrons
#             c(rep(1,4),0,0,rep(1,4)), # HDL
#             c(rep(0,5),1,0,0,0,1), # IDL
#             c(0,0,0,1,0,rep(1,4),0) # (V)LDL
#             )
# 
# # cut off loadings
# w1 <- w$`0.001`[,6:10]
# w2 <- w$`0.001`[,11:14]
# 
# if(all.equal(rownames(w1), colnames(gene))!=T) stop('genes and loadings do not match')
# 
# # Get T
# T1_q <- gene %*% w1
# T1_q_t <- gene_t %*% w1
# 
# T2_q <- gene %*% w2
# T2_q_t <- gene_t %*% w2
# #######################################################################
# # generate Y1 q=23
# y_nonoise <- rbind(T1_q,T1_q_t) %*% t(C1)
# 
# sd_y <- apply(y_nonoise, 2, sd)
# 
# ff <- scale(matrix(rnorm(3465*23,0,1),3465,23)) %*% diag(sd_y) * sqrt(0.5/0.5) # 0.5 is h2_y
# y_all <- y_nonoise + ff
# y <- y_all[1:2000,]
# y_t <- y_all[-(1:2000),]
# 
# ###########################
# # generate Y2 q=10
# yy_nonoise <- rbind(T2_q,T2_q_t) %*% t(C2)
# 
# sd_yy <- apply(yy_nonoise, 2, sd)
# 
# ff <- scale(matrix(rnorm(3465*10,0,1),3465,10)) %*% diag(sd_yy) * sqrt(0.3/0.7) # 0.7 is h2_yy
# yy_all <- yy_nonoise + ff
# yy <- yy_all[1:2000,]
# yy_t <- yy_all[-(1:2000),]
# ######################################
# # for gwas (y1 can read from simu1)
# yy_gwas <- yy
# colnames(yy_gwas) <- paste0('yy',1:10)
# yy_gwas <- cbind(id=rownames(gene), sex=rep(0,2000), yy_gwas) # replace sex here
# write.table(yy_gwas, file=paste0('./fit_dat_simu2/yy_gwas',name_suf), col.names =T, quote = F, row.names = F)
# 
# # for PRSice-2 (y1 can read from simu1)
# yy_prsice <- yy_gwas[,-2]
# colnames(yy_prsice)[1] <- 'IID'
# write.table(yy_prsice, file=paste0('./fit_dat_simu2/yy_prsice', name_suf), col.names =T, quote = F, row.names = F)
# 
# #######################################################################
# # generate Z
# 
# # Z from PCA of true Y1 and Y2
# z1_true <- prcomp(y_nonoise)$x[,1]
# z2_true <- prcomp(yy_nonoise)$x[,1]
# 
# z_true <- 0.5*z1_true + 0.5*z2_true*sd(z1_true)/sd(z2_true) # same contribution from y1 and y2
# ###############################################
# # Z with 90% of heritability
# alpha_z <- (1-0.9)/0.9
# z_all <- z_true + rnorm(3465, 0, sd(z_true)*sqrt(alpha_z))
# z <- z_all[1:2000]
# z_t <- z_all[-(1:2000)]
# # z as liability with prevalence of 0.3
# z_bi <- ifelse(z>quantile(z,0.7),1,0)
# z_bi_t <- ifelse(z_t>quantile(z_t,0.7),1,0)
# 
# # for gwas
# z_gwas <- cbind(id=rownames(gene), sex=rep(0,2000), z=z)
# write.table(z_gwas, file=paste('./fit_dat_simu2/z_gwas',name_suf, sep = ''), col.names =T, quote = F, row.names = F)
# 
# # for PRSice-2
# z_prsice <- z_gwas[,-2]
# colnames(z_prsice)[1] <- 'IID'
# write.table(z_prsice, file=paste('./fit_dat_simu2/z_prsice', name_suf, sep = ''), col.names =T, quote = F, row.names = F)
# ###############################################
# 
# save(T1_q, T1_q_t, w1, T2_q, T2_q_t, w2, y_nonoise, y, y_t, yy_nonoise, yy, yy_t, z_true, z, z_t, z_bi, z_bi_t,
#      file = paste('./fit_dat_simu2/dat_yz', name_suf, '.RData', sep = ''))
#######################################################################
# # # model fitting
# 
# # O2PLS
# fit_o2 <- o2m(gene,yy,4,5,0)
# summary(fit_o2)
# saveRDS(fit_o2, file=paste('./fit_files_simu2/fit_o2', name_suf, '.rds', sep = ''))
# 
# # O2PLS
# fit_o2snp <- o2m(x,yy,4,5,0)
# summary(fit_o2snp)
# saveRDS(fit_o2snp, file=paste('./fit_files_simu2/fit_o2snp', name_suf, '.rds', sep = ''))
# # print('O2PLS SNP: correlation between T and U')
# # cor(fit_o2snp$Tt,fit_o2snp$U)
# 
#PO2PLS
gene <- readRDS(paste('./dat_rep/GeneData_training_',arrID,'.rds', sep = ''))
load(file = paste('./fit_dat_simu2/dat_yz', name_suf, '.RData', sep = ''))

fit_pex <- PO2PLS(gene,yy,4,15,0, steps = 2000, tol=0.01, init_param = 'o2m')
summary(fit_pex)
saveRDS(fit_pex, file=paste('./fit_files_simu2/fit_pex', name_suf, '.rds', sep = ''))

# 
# ##################################
# 
# # GWAS Y2
# # convert.snp.ped('chr1.ped','chr1.map', outfile = 'chr1_ABEL.raw', mapHasHeaderLine = F)
# gwaa_data <- load.gwaa.data(phenofile = paste0('./fit_dat_simu2/yy_gwas',name_suf),
#                             genofile = paste0('./dat_rep/ABEL_',arrID,'.raw'),
#                             force = F, makemap = FALSE, sort = F, id = "id")
# # fix position
# gwaa_data@gtdata@map <- read.table(paste('./dat_rep/chr_all_training_',arrID,'.map', sep = ''), header = F)[,4]
# # gwaa_data@phdata %>% head
# 
# for(yi in 1:10){
#   uni <- qtscore(formula(paste('yy',yi,'~sex',sep = '')), data=gwaa_data,
#                  trait.type = "gaussian", times = 1, quiet = FALSE,
#                  bcast = 10, clambda = TRUE, propPs = 1, details = TRUE)
# 
#   sum_stats <- descriptives.scan(uni, sortby = 'P2df', top = uni@dimresults[1])
#   sum_stats %<>%  mutate(SNP=rownames(sum_stats))
#   # # A2 is effective
#   # effallele(uni)
#   # all.equal(uni@annotation$A2, effallele(uni))
# 
#   write.table(sum_stats, file=paste0('./fit_files_simu2/GWAS_yy',yi, name_suf, '.txt'), row.names = F, col.names = T, quote = F)
# }
# 

# # GWAS Z
# gwaa_data <- load.gwaa.data(phenofile = paste('./fit_dat_simu2/z_gwas',name_suf, sep = ''),
#                             genofile = paste('./dat_rep/ABEL_',arrID,'.raw', sep = ''),
#                             force = F, makemap = FALSE, sort = F, id = "id")
# # fix position
# gwaa_data@gtdata@map <- read.table(paste('./dat_rep/chr_all_training_',arrID,'.map', sep = ''), header = F)[,4]
# # gwaa_data@phdata %>% head
# 
# uni <- qtscore(z~sex, data=gwaa_data,
#                trait.type = "gaussian", times = 1, quiet = FALSE,
#                bcast = 10, clambda = TRUE, propPs = 1, details = TRUE)
# 
# sum_stats <- descriptives.scan(uni, sortby = 'P2df', top = uni@dimresults[1])
# sum_stats %<>%  mutate(SNP=rownames(sum_stats))
# # # A2 is effective
# # effallele(uni)
# # all.equal(uni@annotation$A2, effallele(uni))
# 
# write.table(sum_stats, file=paste0('./fit_files_simu2/GWAS_z', name_suf, '.txt'), row.names = F, col.names = T, quote = F)



print('training R done')